adapter/memQuota/rollingWindow.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memQuota
+
+// Implements a rolling window that allows N units to be allocated per rolling time interval.
+// Time is abstracted in terms of ticks, provided by the caller, decoupling the
+// implementation from real-time, enabling much easier testing and more flexibility.
+type rollingWindow struct {
```

Comment:
at the risk of annoying repetition, I still feel like we should be using something like: https://godoc.org/golang.org/x/time/rate#Limiter here. It uses the token bucket algorithm and should avoid the 11th hour issue you've raised earlier.

Doing so has several advantages (well-tested, widely-used, part of the official supplementary libs, provides a configurable burst). It would also be easier to read and understand (defined in terms that match the user config, no "ticks" to mentally map).

The one thing it doesn't provide is insight into remaining available tokens.  But I'm not clear how that will be used by rate limiting clients in practice. This relates to my other question about what best-effort means for rate limits.